### PR TITLE
Add missing EBCDIC strings for TLS 1.3

### PR DESCRIPTION
**Pull request by opensslonzos-github**
CLA: trivial
Add missing EBCDIC strings
Fix a few places where calling ossl_isdigit does the wrong thing on EBCDIC based systems.
Replaced with ascii_isdigit.
Added one z/OS-only error code used when opening a TTY.

CLA: trivial

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: crypto/rand/rand_unix.c [hunk 1]
```diff
@@ -557,7 +557,62 @@ size_t rand_pool_acquire_entropy(RAND_POOL *pool)
         }
     }
 #   endif
+#if defined(B_ENDIAN) && defined(CHARSET_EBCDIC)
```
Comment (richsalz) [inline crypto/rand/rand_unix.c]: This should be in a separate PR.

Comment (opensslonzos-github) [reply crypto/rand/rand_unix.c]: OK - Will remove

File: crypto/asn1/a_time.c [hunk 2]
```diff
@@ -70,6 +70,11 @@ static void determine_days(struct tm *tm)
     /* Zeller's congruence */
     tm->tm_wday = (d + (13 * m) / 5 + y + y / 4 + c / 4 + 5 * c + 6) % 7;
 }
+static int ascii_isdigit(char inchar) {
+  if (inchar > 0x2F && inchar < 0x3A)
```
Comment (mattcaswell) [inline crypto/asn1/a_time.c]: Style nit: Please use 4 spaces for indentation

File: crypto/asn1/a_time.c [hunk 3]
```diff
@@ -79,6 +84,12 @@ int asn1_time_to_tm(struct tm *tm, const ASN1_TIME *d)
     char *a;
     int n, i, i2, l, o, min_l = 11, strict = 0, end = 6, btz = 5, md;
     struct tm tmp;
+  const char
+    upper_z = 0x5A,
+    num_zero = 0x30,
+    period = 0x2E,
+    hyphen = 0x2D,
+    plus = 0x2B;
```
Comment (mattcaswell) [inline crypto/asn1/a_time.c]: Please put all of these on a single line (breaking at 80 chars)

File: crypto/o_str.c [hunk 4]
```diff
@@ -125,6 +125,40 @@ int OPENSSL_hexchar2int(unsigned char c)
           return 0x0E;
     case 'f': case 'F':
           return 0x0F;
+#ifdef CHARSET_EBCDIC
```
Comment (mattcaswell) [inline crypto/o_str.c]: I'm not sure why you do this here. Would removing the earlier `os_toebcdic` mean that this function just "does the right thing"? Assuming that this function is intended to take a native character set hex digit and convert it to int.

File: crypto/rand/drbg_lib.c [hunk 5]
```diff
@@ -75,8 +75,12 @@ typedef struct drbg_nonce_global_st {
 } DRBG_NONCE_GLOBAL;
 
 /* NIST SP 800-90A DRBG recommends the use of a personalization string. */
+#ifdef CHARSET_EBCDIC
+/* NOTE: If you change DRBG_DEFAULT_PERS_STRING, you will need to change this */
+static const char ossl_pers_string[] = { 0x4f ,0x70 ,0x65 ,0x6e ,0x53 ,0x53 ,0x4c ,0x20 ,0x4e ,0x49 ,0x53 ,0x54 ,0x20 ,0x53 ,0x50 ,0x20 ,0x38 ,0x30 ,0x30 ,0x2d ,0x39 ,0x30 ,0x41 ,0x20 ,0x44 ,0x52 ,0x42 ,0x47 , 0x00};
```
Comment (mattcaswell) [inline crypto/rand/drbg_lib.c]: Wouldn't it be better to change the definition of DRBG_DEFAULT_PERS_STRING instead?

File: crypto/threads_pthread.c [hunk 6]
```diff
@@ -126,7 +126,16 @@ int CRYPTO_THREAD_init_local(CRYPTO_THREAD_LOCAL *key, void (*cleanup)(void *))
...
 {
+#ifndef CHARSET_EBCDIC  
     return pthread_getspecific(*key);
+#else    
+    /*  On z/OS this is 
+    int pthread_getspecific(pthread_key_t key, void **value); 
+    */
+    void *some_value;
+    pthread_getspecific(*key, &some_value);
+    return some_value;
```
Comment (mattcaswell) [inline crypto/threads_pthread.c]: This doesn't seems to have anything to do with EBCDIC character set and has everything to do with z/OS - so this should probably be in a different PR and not use "CHARSET_EBCDIC" to decide what to do (not all EBCDIC systems are z/OS)

File: crypto/ui/ui_openssl.c [hunk 7]
```diff
@@ -443,6 +443,14 @@ static int open_console(UI *ui)
         if (errno == ENODEV)
             is_a_tty = 0;
         else
+#  endif
+#  ifdef ENOSYS
+            /*
+             * z/OS Mainframe can reurn E_NO_SYS (The function is not implemented)
+             */
```
Comment (mattcaswell) [inline crypto/ui/ui_openssl.c]: Please indent the comment to be lined up with the statement below. Again this has nothing to do with EBCDIC and everything to do with z/os so should probably be in a different PR.

File: crypto/x509/x509_vfy.c [hunk 8]
```diff
@@ -1782,13 +1782,19 @@ int X509_cmp_current_time(const ASN1_TIME *ctm)
     return X509_cmp_time(ctm, NULL);
 }
 
+static int ascii_isdigit(char inchar) {
+  if (inchar > 0x2F && inchar < 0x3A)
+    return 1;
+  return 0;
+}
```
Comment (mattcaswell) [inline crypto/x509/x509_vfy.c]: Since this is the second time you've defined this, I suggest moving it to ctype.c instead.

**Comment by opensslonzos-github** [comment]
CLA has been send to legal@opensslfoundation.org.

File: crypto/x509/x509_vfy.c (hunk 8 above)
Comment (opensslonzos-github) [reply crypto/x509/x509_vfy.c]: Move there and added to ctype.h as well

File: crypto/ui/ui_openssl.c (hunk 7 above)
Comment (opensslonzos-github) [reply crypto/ui/ui_openssl.c]: Removed

File: crypto/threads_pthread.c (hunk 6 above)
Comment (opensslonzos-github) [reply crypto/threads_pthread.c]: Removed

File: crypto/rand/drbg_lib.c (hunk 5 above)
Comment (opensslonzos-github) [reply crypto/rand/drbg_lib.c]: The definition used to be in this file (in 1.1.1a code base). I suppose it could be done either way.

File: crypto/asn1/a_time.c (hunk 3 above)
Comment (opensslonzos-github) [reply crypto/asn1/a_time.c]: Done

File: crypto/asn1/a_time.c (hunk 2 above)
Comment (opensslonzos-github) [reply crypto/asn1/a_time.c]: Done

**Comment by mattcaswell** [comment]
Close/reopen to kick CLA bot.

File: crypto/ctype.c [hunk 9]
```diff
@@ -272,3 +272,10 @@ int ossl_toupper(int c)
 {
     return ossl_islower(c) ? c ^ case_change : c;
 }
+
+int ascii_isdigit(const char inchar) {
+    if (inchar > 0x2F && inchar < 0x3A)
```
Comment (paulidale) [inline crypto/ctype.c]: What about `return inchar >= 060 && inchar <= 072;` or equivalent here?

File: crypto/asn1/a_time.c [hunk 10]
```diff
@@ -79,6 +79,7 @@ int asn1_time_to_tm(struct tm *tm, const ASN1_TIME *d)
     char *a;
     int n, i, i2, l, o, min_l = 11, strict = 0, end = 6, btz = 5, md;
     struct tm tmp;
+    const char upper_z = 0x5A, num_zero = 0x30, period = 0x2E, hyphen = 0x2D, plus = 0x2B;
```
Comment (paulidale) [inline crypto/asn1/a_time.c]: What about `minus` instead of `hyphen` since it represents the minus symbol here.

I tend to prefer ASCII in octal, not sure why.

A comment explaining why this isn't `const char upper_z = 'Z', ...` would be appropriate or use a `#ifdef CHARSET_EBCDIC` like is done for the strings elsewhere.

File: crypto/rand/drbg_lib.c [hunk 11]
```diff
@@ -75,8 +75,14 @@ typedef struct drbg_nonce_global_st {
 } DRBG_NONCE_GLOBAL;
 
 /* NIST SP 800-90A DRBG recommends the use of a personalization string. */
+#ifdef CHARSET_EBCDIC
+/* NOTE: If you change DRBG_DEFAULT_PERS_STRING, you will need to change this */
+static const char ossl_pers_string[] = { 0x4f ,0x70 ,0x65 ,0x6e ,0x53 ,0x53 ,
+     0x4c ,0x20 ,0x4e ,0x49 ,0x53 ,0x54 ,0x20 ,0x53 ,0x50 ,0x20 ,0x38 ,0x30 ,
```
Comment (paulidale) [inline crypto/rand/drbg_lib.c]: Swap the spaces and the commas: `0x4f, 0x70, ...`

File: ssl/statem/statem_lib.c [hunk 12]
```diff
@@ -169,9 +169,13 @@ int tls_setup_handshake(SSL *s)
 static int get_cert_verify_tbs_data(SSL *s, unsigned char *tls13tbs,
                                     void **hdata, size_t *hdatalen)
 {
+#ifdef CHARSET_EBCDIC
+    static const char *servercontext = { 0x54 ,0x4c ,0x53 ,0x20 ,0x31 ,0x2e ,0x33 ,0x2c ,0x20 ,0x73 ,0x65 ,0x72 ,0x76 ,0x65 ,0x72 ,0x20 ,0x43 ,0x65 ,0x72 ,0x74 ,0x69 ,0x66 ,0x69 ,0x63 ,0x61 ,0x74 ,0x65 ,0x56 ,0x65 ,0x72 ,0x69 ,0x66 ,0x79, 0x00 };
```
Comment (paulidale) [inline ssl/statem/statem_lib.c]: Same issue with spaces and commas here too.

The space/comma thing isn't consistent.

File: crypto/ctype.c (hunk 9 above)
Comment (opensslonzos-github) [reply crypto/ctype.c]: I would really like to keep these in hex as all the other ASCII/EBCDIC charset issues are expressed that way.

File: crypto/asn1/a_time.c (hunk 10 above)
Comment (opensslonzos-github) [reply crypto/asn1/a_time.c]: Adding a comment and changing the name to minus.

File: ssl/statem/statem_lib.c (hunk 12 above)
Comment (opensslonzos-github) [reply ssl/statem/statem_lib.c]: Fixing

File: crypto/rand/drbg_lib.c (hunk 11 above)
Comment (opensslonzos-github) [reply crypto/rand/drbg_lib.c]: Sorry about that

File: crypto/o_str.c (hunk 4 above)
Comment (mattcaswell) [reply crypto/o_str.c]: This comment seems to be still unresolved.

File: crypto/rand/drbg_lib.c [hunk 13]
```diff
@@ -75,8 +75,14 @@ typedef struct drbg_nonce_global_st {
 } DRBG_NONCE_GLOBAL;
 
 /* NIST SP 800-90A DRBG recommends the use of a personalization string. */
+#ifdef CHARSET_EBCDIC
+/* NOTE: If you change DRBG_DEFAULT_PERS_STRING, you will need to change this */
```
Comment (mattcaswell) [inline crypto/rand/drbg_lib.c]: IMO it would be better to make this change in rand_lcl.h, i.e.

    #ifdef CHARSET_EBCDIC
    # define DRBG_DEFAULT_PERS_STRING { 0x4f .... }
    #else
    # define DRBG_DEFAULT_PERS_STRING "OpenSSL NIST SP 800-90A DRBG"
    #endif

Comment (opensslonzos-github) [reply crypto/rand/drbg_lib.c]: Moved as you suggested

File: crypto/o_str.c (hunk 4 above)
Comment (opensslonzos-github) [reply crypto/o_str.c]: Removed

**Comment by mattcaswell** [comment]
I squashed the commits (and made some minor edits to the original commit message) and pushed to master and 1.1.1. There were some trivial conflicts during the cherry-pick that were easily resolved.
